src/cpuinfo_ppc.c

```diff
@@ -0,0 +1,359 @@
+    {{PPC_FEATURE_ARCH_2_06, 0}, "arch_2_06", &set_arch206},
+    {{PPC_FEATURE_HAS_VSX, 0}, "vsx", &set_vsx},
+    {{PPC_FEATURE_PSERIES_PERFMON_COMPAT, 0}, "archpmu", &set_pseries_perfmon_compat},
+    {{PPC_FEATURE_TRUE_LE, 0}, "true_le", &set_truele},
+    {{PPC_FEATURE_PPC_LE, 0}, "ppcle", &set_ppcle},
+    {{0, PPC_FEATURE2_ARCH_2_07}, "arch_2_07", &set_arch207},
+    {{0, PPC_FEATURE2_HTM}, "htm", &set_htm},
+    {{0, PPC_FEATURE2_DSCR}, "dscr", &set_dscr},
+    {{0, PPC_FEATURE2_EBB}, "ebb", &set_ebb},
+    {{0, PPC_FEATURE2_ISEL}, "isel", &set_isel},
+    {{0, PPC_FEATURE2_TAR}, "tar", &set_tar},
+    {{0, PPC_FEATURE2_VEC_CRYPTO}, "vcrypto", &set_vcrypto},
+    {{0, PPC_FEATURE2_HTM_NOSC}, "htm-nosc", &set_htm_nosc},
+    {{0, PPC_FEATURE2_ARCH_3_00}, "arch_3_00", &set_arch300},
+    {{0, PPC_FEATURE2_HAS_IEEE128}, "ieee128", &set_ieee128},
+    {{0, PPC_FEATURE2_DARN}, "darn", &set_darn},
+    {{0, PPC_FEATURE2_SCV}, "scv", &set_scv},
+    {{0, PPC_FEATURE2_HTM_NO_SUSPEND}, "htm-no-suspend", &set_htm_no_suspend},
+};
+static const size_t kConfigsSize = sizeof(kConfigs) / sizeof(CapabilityConfig);
+
+static bool HandlePPCLine(const LineResult result,
+                              PPCPlatformStrings* const strings) {
+  StringView line = result.line;
+  StringView key, value;
+  if (GetAttributeKeyValue(line, &key, &value)) {
+    if (HasWord(key, "platform")) {
+      CopyString(value, strings->platform, value.size + 1);
+    } else if (IsEquals(key, str("model"))) {
+      CopyString(value, strings->model, value.size + 1);
+    } else if (IsEquals(key, str("machine"))) {
+      CopyString(value, strings->machine, value.size + 1);
+    } else if (IsEquals(key, str("cpu"))) {
+      CopyString(value, strings->cpu, value.size + 1);
+    }
+  }
+  return !result.eof;
+}
+
+static void FillProcCpuInfoData(PPCPlatformStrings* const strings) {
+  const int fd = OpenFile("/proc/cpuinfo");
+  if (fd >= 0) {
+    StackLineReader reader;
+    StackLineReader_Initialize(&reader, fd);
+    for (;;) {
+      if (!HandlePPCLine(StackLineReader_NextLine(&reader), strings)) {
+        break;
+      }
+    }
+    CloseFile(fd);
+  }
+}
+
+
+static const PPCInfo kEmptyPPCInfo;
+
+PPCInfo GetPPCInfo(void) {
+  /*
+   * On Power feature flags aren't currently in cpuinfo so we only look at
+   * the auxilary vector
+   */
+  PPCInfo info = kEmptyPPCInfo;
+
+  OverrideFromHwCaps(kConfigsSize, kConfigs, GetHardwareCapabilities(),
+                     &info.features);
+  return info;
+}
+
+
+static char platform[MAX_LINE_SIZE];
+static char model[MAX_LINE_SIZE];
+static char machine[MAX_LINE_SIZE];
+static char cpu[MAX_LINE_SIZE];
+static const PPCPlatformStrings kEmptyPPCPlatformStrings = {
+	.platform = platform, .model = model, .machine = machine, .cpu = cpu};
+
+PPCPlatformStrings GetPPCPlatformStrings(void) {
+  PPCPlatformStrings strings = kEmptyPPCPlatformStrings;
+
+  FillProcCpuInfoData(&strings);
```

Comment:
This is a bit gross sorry. 

At first I did want to avoid dynamic memory allocation, but then I realised that most of the information needed is can be deduced from the feature flags*. Hence not including the `PPCPlatformStrings` struct in the  `PPCInfo` struct to keep `GetPPCInfo()` memcpy/malloc free by avoiding strings.

I was going to only parse `/proc/cpuinfo` once in `GetPPCPlatformStrings` and have a static 'PPCPlatformStrings' struct that would be copied everytime  `GetPPCPlatformStrings` was called. But then I figured I should talk to you about that, and didn't tidy up properly :woman_facepalming:  

Do you think it makes sense to only parse `/proc/cpuinfo` once in `GetPPCPlatformStrings`? Or should I leave that behaviour as is because a process shouldn't repeatedly call it?  

*The only information that I think you could gain from the `strings` struct that you can't from the flags would be the actual hardware architecture if your process was running in compat mode (ie running with POWER7 instructions on a POWER8 architecture).

Reply:
I actually discussed this internally when designing the library. We decided that `GetXXXInfo()` should do all the work always from scratch and not try to reuse data. This removes the problem of thread synchronization entirely. On the performance side this is not a big deal and the client can always cache the result if need be.

There's no need for dynamic allocation, just return a struct containing an array of char, it will be copied for you when the struct is returned.